clarity_ext/service/dilution/service.py

```diff
@@ -532,6 +549,90 @@ def output_position_sort_key(transfer):
                 transfer.source_slot.index,
                 transfer.source_location.index_down_first)
 
+    @staticmethod
+    def container_sort_key(container):
+        """
+        Separates text and numbers in the container name
+        """
+        def conv_int(s):
+            if s.isdigit():
+                return int(s)
+            else:
+                return str(s)
+
+        def conv_lower(s):
+            return s.lower()
+
+        def zip_order(a_list):
+            # the list starting with an empty string should be sorted first into zip()
+            return len(a_list[0])
+
+        strings_in_name = re.split('[-_]*\d+[-_]*', container.name)
+        strings_in_name = map(conv_lower, strings_in_name)
+        numbers_in_name = re.split('[-_]*\D+[-_]*', container.name)
+        zip_arg = sorted([numbers_in_name, strings_in_name], key=zip_order)
+        zipped = zip(*zip_arg)
+        flatlist = list(chain(*zipped))
+        sortlist_of_strings = [item for item in flatlist if len(item) > 0]
+        sortlist = [container.sort_weight, not container.is_temporary] + map(conv_int, sortlist_of_strings)
+        return tuple(sortlist)
+
+
+class TubeRackPositioner:
+    """
+    Positions tubes in tube racks.
+    Holds a state on how many tube racks is needed, and
+    a mapping between tubes and positions in tube racks
+    """
+    TUBE_RACK_START_ID = 1211111111
+
+    def __init__(self, plate_size):
+        self.tube_racks = list()
+        self.tube_rack_id_counter = 0
+        self.tube_counter = 1
+        self.size = plate_size
+        self.number_of_wells = self.size.height * self.size.width
+        self.current_well_pos = None
+
+    def add(self, well):
+        """
+        Place the tube for the input well into a tube rack
+        :param well: This is the single well for a tube
+        :return:
+        """
+        position_index = (self.tube_counter - 1) % self.number_of_wells
+        tube_rack_index = (self.tube_counter - 1) // self.number_of_wells
+        self.tube_counter += 1
+        if len(self.tube_racks) < tube_rack_index + 1:
+            self._create_new_tube_rack()
+        row_ind, col_ind = self._convert_to_coordinates(position_index)
+        self.current_well_pos = ContainerPosition(row_ind + 1, col_ind + 1)
+
+        self.tube_racks[-1].set_well(well_pos=self.current_well_pos, artifact=well.artifact)
+
+    @property
+    def last_populated_well(self):
+        return self.tube_racks[-1].wells[self.current_well_pos]
+
+    def _convert_to_coordinates(self, position_index):
+        rowind = position_index % self.size.height
+        colind = position_index // self.size.height
+        return rowind, colind
+
+    def _convert_to_position(self, row_ind, col_ind):
```

Comment:
I wonder if this function should return a string or `ContainerPosition`. It's not used at the moment, so difficult to be sure.